packages/jsii-pacmak/lib/targets/python/type-name.ts

```diff
@@ -0,0 +1,354 @@
+   * indicates a need to import the module fully ("import <name>") instead of
+   * doing a piecemeal import ("from <name> import <item>").
+   */
+  readonly [sourcePackage: string]: ReadonlySet<string>;
+}
+
+/**
+ * The context in which a PythonType is being considered.
+ */
+export interface NamingContext {
+  /** The assembly in which the PythonType is expressed. */
+  readonly assembly: Assembly;
+
+  /** The submodule of the assembly in which the PythonType is expressed (could be the module root) */
+  readonly submodule: string;
+
+  /**
+   * The declaration is made in the context of a type annotation (so it can be quoted)
+   *
+   * @default true
+   */
+  readonly typeAnnotation?: boolean;
+
+  /**
+   * The nesting scope in which the PythonType is expressed (if any)
+   *
+   * @default - none
+   */
+  readonly nestingScope?: string | undefined;
+
+  /**
+   * Disables generating typing.Optional wrappers
+   * @default false
+   * @internal
+   */
+  readonly ignoreOptional?: boolean;
+}
+
+export function toTypeName(ref?: OptionalValue | TypeReference): TypeName {
+  if (ref == null) {
+    return Primitive.NONE;
+  }
+
+  const type = isOptionalValue(ref) ? ref.type : ref;
+  const optional = isOptionalValue(ref) && ref.optional;
+
+  let result: TypeName = Primitive.ANY;
+
+  if (isPrimitiveTypeReference(type)) {
+    result = Primitive.of(type);
+  } else if (isCollectionTypeReference(type)) {
+    const elt = toTypeName(type.collection.elementtype);
+    if (type.collection.kind === CollectionKind.Array) {
+      result = new List(elt);
+    } else {
+      result = new Dict(elt);
+    }
+  } else if (isUnionTypeReference(type)) {
+    result = new Union(type.union.types.map(toTypeName));
+  } else if (isNamedTypeReference(type)) {
+    result = new UserType(type.fqn);
+  }
+
+  return optional ? new Optional(result) : result;
+}
+
+export function mergePythonImports(...pythonImports: readonly PythonImports[]): PythonImports {
+  const result: Record<string, Set<string>> = {};
+  for (const bag of pythonImports) {
+    for (const [packageName, items] of Object.entries(bag)) {
+      if (!(packageName in result)) {
+        result[packageName] = new Set();
+      }
+      for (const item of items) {
+        result[packageName].add(item);
+      }
+    }
+  }
+  return result;
+}
+
+function isOptionalValue(type: OptionalValue | TypeReference): type is OptionalValue {
+  return (type as unknown as OptionalValue).type != null;
+}
+
+class Dict implements TypeName {
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #element: TypeName;
+
+  public constructor(element: TypeName) {
+    this.#element = element;
+  }
+
+  public pythonType(context: NamingContext) {
+    return `typing.Mapping[str,${this.#element.pythonType(context)}]`;
+  }
+
+  public requiredImports(context: NamingContext) {
+    return mergePythonImports(this.#element.requiredImports(context), { typing: new Set(['']) });
```

Comment:
why is this needed? It's very unclear.

Reply:
Well because I need to import:
1. What needs to be imported for the element type
2. `typing` (so I can have `typing.Mapping` in scope)

Reply:
Seems like typing is always needed, why specify every time?

Reply:
Typing isn't needed for user-defined types (named references). It's often needed, not always.